### PR TITLE
Adding support to multi-consumers on BurstCostExecutorsBenchmark

### PR DESCRIPTION
**Pull request by franz1981**
Motivation:

Sometimes is needed to measure the performance of multi-consumers
scenarios with bursty load.

Modifications:

Switching to AverageTime avoid to rely on OS nanoTime granularity.
Removing the spinning event executor, because it is not representative
of anything used in the library and is not taking into accounts the
multi-consumers case.
Modified DefaultEventExecutorChooserFactory by:
- including the single executor case and avoid array usage
- improving the other 2 cases by avoiding multiple loads of fields

Result:

The benchmark can be used to measure interactions between executors
in multi-consumer scenarios.

### CODE REVIEW

**Comment by franz1981** [comment]
I've modified this benchmark in order to use it to test the cost of https://github.com/netty/netty/pull/9025

File: common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java [hunk 1]
```diff
@@ -43,6 +46,19 @@ private static boolean isPowerOfTwo(int val) {
         return (val & -val) == val;
     }
 
+    private static final class SingleEventExecutorChooser implements EventExecutorChooser {
```
Comment (franz1981) [inline common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java]: @normanmaurer @njhill 
On Netty 5 these 3 classes could be collapsed into lambdas in order to let the JVM trust non static final fields (see http://hg.openjdk.java.net/jdk/jdk/file/2c1af559e922/src/hotspot/share/ci/ciField.cpp#l203) eg executors array. 
There is a good chance that it could be just implemented with:
```java
@Override
    public EventExecutorChooser newChooser(final EventExecutor[] executors) {
        final int executorsLength = executors.length;
        final AtomicInteger idx = new AtomicInteger();
        if (executorsLength > 0) {
            return () -> {
                return executors[Math.abs(idx.getAndIncrement() % executorsLength)];
            };
        } else {
            final EventExecutor executor = executors[0];
            return () -> {
                return executor;
            }
        }
    }
```

File: common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java [hunk 2]
```diff
@@ -53,7 +69,10 @@ private static boolean isPowerOfTwo(int val) {
 
         @Override
         public EventExecutor next() {
-            return executors[idx.getAndIncrement() & executors.length - 1];
+            final AtomicInteger idx = this.idx;
+            final EventExecutor[] executors = this.executors;
+            final int executorsMask = executors.length - 1;
+            return executors[idx.getAndIncrement() & executorsMask];
```
[the remark this answers is not in the document]
Comment (franz1981) [reply common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java]: Exactly, a similar pattern is been used in the many concurrent primitives of `java.util.concurrent.*` classes; copied from `ForkJoinPool` code:
```
     * Memory ordering relies mainly on Unsafe intrinsics that carry
     * the further responsibility of explicitly performing null- and
     * bounds- checks otherwise carried out implicitly by JVMs.  This
     * can be awkward and ugly, but also reflects the need to control
     * outcomes across the unusual cases that arise in very racy code
     * with very few invariants. So these explicit checks would exist
     * in some form anyway.  All fields are read into locals before
     * use, and null-checked if they are references.  This is usually
     * done in a "C"-like style of listing declarations at the heads
     * of methods or blocks, and using inline assignments on first
     * encounter.  Array bounds-checks are usually performed by
     * masking with array.length-1, which relies on the invariant that
     * these arrays are created with positive lengths, which is itself
     * paranoically checked.
```
TLDR; primitives like `getAndIncrement()` can force instance fields to be read several times and bounds checks to happen more then necessary: on hot paths it doesn't cost nothing (if not ugliness of code) to use this declaration style.

File: microbench/src/main/java/io/netty/microbench/concurrent/BurstCostExecutorsBenchmark.java [hunk 3]
```diff
@@ -176,60 +52,48 @@ public void execute(Runnable command) {
...
     private int burstLength;
-    @Param({ "spinning", "epollEventLoop", "nioEventLoop", "defaultEventExecutor", "juc", "kqueueEventLoop" })
+    @Param({ "epollEventLoop", "nioEventLoop", "defaultEventExecutor", "juc", "kqueueEventLoop" })
     private String executorType;
     @Param({ "0", "10" })
     private int work;
+    @Param({ "1", "2" })
+    private int consumers;
 
     private ExecutorService executor;
```
[the remark this answers is not in the document]
Comment (franz1981) [reply microbench/src/main/java/io/netty/microbench/concurrent/BurstCostExecutorsBenchmark.java]: The benchmark support using common Java `ExecutorServices` too

File: microbench/src/main/java/io/netty/microbench/concurrent/BurstCostExecutorsBenchmark.java [hunk 4]
```diff
@@ -297,26 +183,19 @@ public int spinWaitCompletionOf(int value) {
...
+    @BenchmarkMode(Mode.AverageTime)
     @Threads(2)
     public int test2Producers(final PerThreadState state) {
         return executeBurst(state);
     }
 
-    @Benchmark
-    @BenchmarkMode(Mode.SampleTime)
-    @Threads(3)
-    public int test3Producers(final PerThreadState state) {
```
[the remark this answers is not in the document]
Comment (franz1981) [reply microbench/src/main/java/io/netty/microbench/concurrent/BurstCostExecutorsBenchmark.java]: Just matter of math on common machines: thoeretically is not a good idea to run benchs on a dev machine, but sometime it could happen; many dev machines has just 4 physical cores ie a run with 2 consumers and 2 producers is already maximizing the machine. Using 5 cores (3 producers + 2 consumers) could just produce werid results...

File: common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java [hunk 5]
```diff
@@ -32,6 +32,9 @@ private DefaultEventExecutorChooserFactory() { }
     @SuppressWarnings("unchecked")
     @Override
     public EventExecutorChooser newChooser(EventExecutor[] executors) {
+        if (executors.length == 1) {
```
[the remark this answers is not in the document]
Comment (franz1981) [reply common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java]: Yep, my suggestion for Netty 5 would be to just make it simpler using lambdas, given that the JVM trust captured final fields and this could make the code much simpler without perf losses

**Comment by franz1981** [comment]
@njhill Answers inline

> Why is a completed counter needed? Could you just submit burstLength - 1 workTask runnables that don't touch that field at all followed by one that flips it (i.e. could just be boolean which the spin-loop waits for)?

It is due to ordering reasons: any `execute` can be consumed by different threads in different orders and the final execution could be executed before the other ones. 
We are interested into waiting that all the submitted runnables are being executed, instead of just the last one. With a single consumer this can't happen.
The same error seems present here: https://github.com/JCTools/JCTools/blob/master/jctools-benchmarks/src/main/java/org/jctools/jmh/latency/MpqBurstCost.java#L75
But maybe the mentioned benchmark isn't measuring the same thing of this one (sorry to bother @nitsanw , but I'm sure he has some information/hint about it).

> Also I may have a naive understanding of this but despite the lazy set being "safe" here, might it not actually slow down the bench since there's no "rush" for the spin loop to observe the change?

The lazy name was probably the worst name ever and the worst javadoc in the history of the concurrent util classes IMO: the actual meaning of such `lazy` is `ordered` or better `writeRelease` ie it will be perceived immediately by a reading thread exactly like a `volatile` store, just with different effects on the writer for both order of subsequent operations and perf wise (but is an implementaion detail).

>  If it was decided not to use the lazy increment then a simpler impl could be to do a submit rather than execute for the final task and wait on the returned future... not sure if it might be less efficient in some other way though?

It would be interesting to do it as a separate benchmark IMO, for many important reasons:

1. `submit` is more likely to create some "noise" due to garbage produced that would interefer with the evaluation of submission cost (and duty cycle wakeup + execution)
2. awaiting using *common* java primitives (`CountdownLatch ` or similar) is slow  (eg see https://issues.apache.org/jira/browse/QPIDJMS-396) and could create noise that I wasn't interested to measure here

Nonetheless it could be an interesting benchmark, just I'm not sure it is a common user case (or it is? Any opinion is wellcome!), that's why I haven't added it yet.
As you have noticed that benchmark is not as "clean" as it could be: when `consumers>1` the consumer duty cycles interefere with each others while incrementing an atomic int, that's not perfectly fine, but I haven't found any other way to track that many unordered executions would be entirely executed...

File: common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (njhill) [reply common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java]: I guess I don't follow the direct equivalence between this what's described there which appears to relate to the need for explicit bounds/null checks due to use of `Unsafe`, I am probably still missing the subtlety! Is it that the non-plain `getAndIncrement` invocation could happen between reads of the `executors` field? I would have thought the most that could help is a local var for that once since it's the only one read more than once?

[the remark this answers is not in the document]
Comment (franz1981) [reply common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java]: I've probably misjudged the code I've seen and I can revert this change in this specific case. The original idea is to control how many times  "executors" has been fetched and null checked (in this case, by the JVM) due to the ordering imposed by getAndIncrement.
I see now that "executors" will be used right after getAndIncrement ie it shouldn't fetched twice.

**Comment by nitsanw** [comment]
@franz1981 You are right, there's an issue in that benchmark for multiple consumers. Good catch!

**Comment by njhill** [comment]
Thanks @franz1981 
> It is due to ordering reasons: any execute can be consumed by different threads in different orders and the last could be executed before the previous ones.
I suppose we are interested into waiting that all the submitted runnables are being executed, instead of just the last one. With a single consumer this can't happen.

Right, my original thought was prior to this change to include multi-consumers. Maybe it would still be worthwhile to do this for the single consumer case? I have a feeling that's the case we care about most from a perf mesurement/optimization since the multi-consumer aspect is kind of a layer on top and includes eventloop selection etc. I don't think it's common that tasks are submitted to an ELG, normally it's in the context of a single EL, i.e. they weren't really designed to be efficient general-purpose MPMC threadpools even though it still "works".

Re the submit/future idea, was just an idea for simplifying the benchmark logic but agree that's it's not worth pursuing given your answer to my lazySet question :)

**Comment by normanmaurer** [comment]
So @njhill @franz1981 what is the status here ?

**Comment by franz1981** [comment]
@normanmaurer 
Sorry, super busy period :(
So, I would rebase this and create another benchmark ie MultiBurstCostExecutorsBenchmark, because I see that the version I've put here is too "different" from the single producer case, but I could be fine to close this one too if on Netty the submission of tasks from multiple producers is not an interesting case.
The motivation of benchmarking should be to spot regressions/understand the behaviour of interesting cases, so...wdyt?

**Comment by normanmaurer** [comment]
@franz1981 works for me...

**Comment by normanmaurer** [comment]
@franz1981 ping

**Comment by franz1981** [comment]
@normanmaurer I see 2 issues with this benchmark:

- multi-consumer scenario is not a common scenario AFAIK: it makes this PR not so attractive
- awaking/sleeping logic of the event loop (as demonstrated recently) are not that reproducible nor so easy to be stressed using a black-box approach so I need to re-think about the purpose of the single consumer version of this same benchmark too

TLDR: I would like to create a benchmark that effectively capture our common/critical cases on Netty, so I can close this a open a discussion about what we really want to measure, wdyt?

**Comment by normanmaurer** [comment]
👍
